app/state/services.py

```diff
@@ -137,23 +131,66 @@ def get_ip(self, headers: Mapping[str, str]) -> IPAddress:
         return ip
 
 
-def fetch_geoloc_db(ip: IPAddress) -> Geolocation:
-    """Fetch geolocation data based on ip (using local db)."""
-    assert geoloc_db is not None
+async def fetch_geoloc(
+    ip: IPAddress,
+    headers: Optional[Mapping[str, str]] = None,
+) -> Optional[Geolocation]:
+    """Fetch geolocation data based on ip."""
+    geoloc = fetch_geoloc_cloudflare(ip, headers)
+
+    if geoloc is None:
+        geoloc = fetch_geoloc_nginx(ip, headers)
+
+    if geoloc is None:
+        geoloc = await fetch_geoloc_web(ip)
+
+    return geoloc
+
 
-    res = geoloc_db.city(ip)
+def fetch_geoloc_cloudflare(
+    ip: IPAddress,
+    headers: Mapping[str, str],
+) -> Optional[Geolocation]:
+    """Fetch geolocation data based on ip (using cloudflare headers)."""
+    country_code = headers.get("CF-IPCountry", "xx")
 
-    if res.country.iso_code is not None:
-        acronym = res.country.iso_code.lower()
-    else:
-        acronym = "XX"
+    if country_code is None:
```

Comment:
country_code will never be none. check if CF-IPCountry, CF-IPLatitude or CF-IPLongitude are missing and return. otherwise go on with the code (drop the gets as well